File: mage/src/main/java/mil/nga/giat/mage/profile/ProfilePictureViewerActivity.java
package mil.nga.giat.mage.profile;

import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.resource.drawable.GlideDrawable;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.Target;

import org.apache.commons.lang3.StringUtils;

import java.io.File;
import java.util.Collections;

import mil.nga.giat.mage.R;
import mil.nga.giat.mage.sdk.datastore.user.User;
import mil.nga.giat.mage.sdk.datastore.user.UserHelper;
import mil.nga.giat.mage.sdk.datastore.user.UserLocal;
import mil.nga.giat.mage.sdk.exceptions.UserException;
import mil.nga.giat.mage.sdk.fetch.DownloadImageTask;

public class ProfilePictureViewerActivity extends AppCompatActivity {
	private static final String LOG_NAME = ProfilePictureViewerActivity.class.getName();
	
	public final static String USER_ID = "USER_ID";

	private static final int MAX_DOWNLOAD_ATTEMPS = 1;

	ImageView imageView;
	View progress;
	int attempts;
	
	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.attachment_viewer);
		getSupportActionBar().setDisplayHomeAsUpEnabled(true);

		Long userID = getIntent().getLongExtra(USER_ID, -1);

		imageView = (ImageView)findViewById(R.id.image);
		progress = findViewById(R.id.progress);

		attempts = 0;

		if (userID >= 0) {
			try {
				final User user = UserHelper.getInstance(getApplicationContext()).read(userID);
				final UserLocal userLocal = user.getUserLocal();
				this.setTitle(user.getDisplayName());

				String localAvatarPath = userLocal.getLocalAvatarPath();

				if (StringUtils.isNotBlank(localAvatarPath)) {
					setProfilePicture(user);
				} else if (user.getAvatarUrl() != null) {
					downloadProfilePicture(user);
				} else {
					progress.setVisibility(View.GONE);
					findViewById(R.id.no_content).setVisibility(View.VISIBLE);
				}
			} catch(Exception e) {
				Log.e(LOG_NAME, "Could not set title.", e);
			}
		}
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		switch (item.getItemId()) {
			case android.R.id.home:
				this.finish();
				return true;
		}
		return true;
	}

	private void downloadProfilePicture(final User user) {
		if (attempts >= MAX_DOWNLOAD_ATTEMPS) {
			showErrorDialog();
			return;
		}

		attempts++;

		if (user.getAvatarUrl() != null) {
			new DownloadImageTask(getApplicationContext(), Collections.singletonList(user), DownloadImageTask.ImageType.AVATAR, true, new DownloadImageTask.OnImageDownloadListener() {
				@Override
				public void complete() {
					try {
						User updatedUser = UserHelper.getInstance(getApplicationContext()).read(user.getId());
						String avatarPath = updatedUser.getUserLocal().getLocalAvatarPath();
						if (avatarPath != null) {
							setProfilePicture(updatedUser);
						} else {
							showErrorDialog();
						}
					} catch (UserException e) {
						e.printStackTrace();
					}

				}
			}).execute();
		}
	}

	private void setProfilePicture(final User user) {
		Glide.with(this)
				.load(new File(user.getUserLocal().getLocalAvatarPath()))
				.listener(new RequestListener<File, GlideDrawable>() {
					@Override
					public boolean onException(Exception e, File model, Target<GlideDrawable> target, boolean isFirstResource) {
						downloadProfilePicture(user);
						return false;
					}

					@Override
					public boolean onResourceReady(GlideDrawable resource, File model, Target<GlideDrawable> target, boolean isFromMemoryCache, boolean isFirstResource) {
						progress.setVisibility(View.GONE);
						return false;
					}
				})
				.into(imageView);
	}

	private void showErrorDialog() {
		new AlertDialog.Builder(ProfilePictureViewerActivity.this, R.style.AppCompatAlertDialogStyle)
				.setTitle("Error Downloading Avatar")
				.setMessage("MAGE could not download this users avatar.  Please try again later.")
				.setPositiveButton(android.R.string.ok, null)
				.create()
				.show();
	}
}

File: mage/src/main/java/mil/nga/giat/mage/profile/ProfilePictureViewerActivity.kt
package mil.nga.giat.mage.profile

import android.content.Context
import android.content.Intent
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import com.bumptech.glide.request.target.Target

import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.request.RequestListener
import mil.nga.giat.mage.R
import mil.nga.giat.mage.databinding.AttachmentViewerBinding
import mil.nga.giat.mage.glide.GlideApp
import mil.nga.giat.mage.glide.model.Avatar
import mil.nga.giat.mage.sdk.datastore.user.User
import mil.nga.giat.mage.sdk.datastore.user.UserHelper

class ProfilePictureViewerActivity : AppCompatActivity() {

    companion object {
        private val LOG_NAME = ProfilePictureViewerActivity::class.java.name

        private const val USER_ID_EXTRA = "USER_ID"

        fun intent(context: Context, user: User): Intent {
            val intent = Intent(context, ProfilePictureViewerActivity::class.java)
            intent.putExtra(USER_ID_EXTRA, user.id)
            return intent
        }
    }

    private lateinit var binding: AttachmentViewerBinding

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        require(intent.hasExtra(USER_ID_EXTRA)) {"USER_ID_EXTRA is required to launch ProfilePictureViewerActivity"}

        binding = AttachmentViewerBinding.inflate(layoutInflater)
        setContentView(binding.root)

        supportActionBar?.let {
            it.setDisplayHomeAsUpEnabled(true)
            it.setDisplayShowHomeEnabled(true)
        }

        binding.progress.visibility = View.VISIBLE

        try {
            val userID = intent.getLongExtra(USER_ID_EXTRA, -1)
            val user = UserHelper.getInstance(applicationContext).read(userID)
            this.title = user.displayName

            GlideApp.with(this)
                .load(Avatar.forUser(user))
                .listener(object : RequestListener<Drawable> {
                    override fun onLoadFailed(e: GlideException?, model: Any, target: Target<Drawable>, isFirstResource: Boolean): Boolean {
                        binding.progress.visibility = View.GONE
                        binding.noContent.visibility = View.VISIBLE
                        return false
                    }

                    override fun onResourceReady(resource: Drawable, model: Any, target: Target<Drawable>, dataSource: DataSource, isFirstResource: Boolean): Boolean {
                        binding.progress.visibility = View.GONE
                        return false
                    }
                })
                .into(binding.image)
        } catch (e: Exception) {
            Log.e(LOG_NAME, "Could not set title.", e)
            binding.progress.visibility = View.GONE
            showErrorDialog()
        }
    }

    override fun onSupportNavigateUp(): Boolean {
        finish()
        return true
    }

    private fun showErrorDialog() {
        AlertDialog.Builder(this, R.style.AppCompatAlertDialogStyle)
            .setTitle("Error Downloading Avatar")
            .setMessage("MAGE could not download this users avatar.  Please try again later.")
            .setPositiveButton(android.R.string.ok) { _, _ ->  finish()}
            .create()
            .show()
    }
}
